[Perf Experiment][ConstraintGraph] Gather constraints improvements

xedin opened the pull request:
- Use set to gather constraints for type variables to avoid duplicates
- Add filtering to `gatherConstraints` per type variable

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintGraph] Use set to gather constraints for type variables

Since it's possible to find the same constraint through two different
but equivalent type variables, let's use a set to store constraints
instead of a vector to avoid processing the same constraint multiple
times.

xedin committed:
[ConstraintGraph] Add filtering to `gatherConstraints` per type variable

Most of the use-cases of `gatherConstraints` require filtering
at least based on the constraint kind that caller is interested in,
so instead of returning unrelated results and asking caller to
filter separately, let's add that functionality directly to
`gatherConstraints`.

rudkx commented on lib/Sema/CSBindings.cpp:
```diff
@@ -361,8 +361,7 @@ ConstraintSystem::getPotentialBindings(TypeVariableType *typeVar) {
   // Gather the constraints associated with this type variable.
-  SmallVector<Constraint *, 8> constraints;
-  llvm::SmallPtrSet<Constraint *, 4> visitedConstraints;
+  SmallPtrSet<Constraint *, 8> constraints;
```
Sets are unordered collections and as a result you have no stable order when iterating over them. This is especially bad with sets of pointers since you're almost certainly guaranteed change in order from run to run.

xedin replied:
I can switch it to `SmallSetVector` which should have insertion order iteration. Will try that next.

rudkx replied:
Oh it looks like `C++`'s `std::set` is guaranteed to use `std::less` and you're guaranteed an iteration order, but again with pointers as the elements you will not be guaranteed to iterate over the pointed-to elements in the same order from run to run.

xedin committed:
[ConstraintGraph] Change `gatherConstraints` to take `SetVector`

For stable iteration order, let's switch from `SmallPtrSet`
to `SetVector` which ensures insertion order iteration.

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please smoke test compiler performance

DougGregor commented on lib/Sema/CSDiag.cpp:
```diff
@@ -8727,29 +8728,28 @@ bool FailureDiagnosis::diagnoseArchetypeAmbiguity() {
+          tv, constraints, ConstraintGraph::GatheringKind::EquivalenceClass,
+          [&](Constraint *constraint) -> bool {
+            // We are not interested in ConformsTo constraints because
+            // such constraints specify restrictions on the archetypes
```
I'd rather not say "archetypes" here; I think the reason we don't care about `conformsTo` constraints is that we can't derive any concrete type information from them.

xedin replied:
Sounds good! It's an existing comment I've moved here but I'll fix it up if perf test is green...

DougGregor reviewed (approved):
Definitely cleaner!

xedin committed:
[Diagnostics] NFC: Clarify why we don't care about  while diagnosing archetype errors

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility
